File: dp/coinChange.js
// 换硬币(返回可以凑成总金额所需的, 最少的硬币个数)
function coinChange(coins, amount) {
    let dp = new Array(amount + 1).fill(Infinity)
    dp[0] = 0
    for (let i = 1; i <= amount; i++) {
        for (let coin of coins) {
            if (i >= coin) {
                // 金额必须大于硬币额度
                dp[i] = Math.min(dp[i], dp[i - coin] + 1);
            }
        }
    }
    return dp[amount] === Infinity ? -1 : dp[amount]
}

/**
 * 思路
 * min(F(amount-coins[0])+1, F(amount-coins[n])+1)
 */

// dp[ i ] 表示最小的兑换硬币个数
// dp[ i ] = Min(dp[ amount - coins[n] ] + 1, dp[ i ])
const coins = [ 1, 2, 4 ];
const amount = 11;
const ret = coinChange(coins, amount);
console.log(ret);

File: dp/coinChange.go
package main

import "math"

// 给你一个整数数组 coins ，表示不同面额的硬币；以及一个整数 amount ，表示总金额。
// 计算并返回可以凑成总金额所需的 最少的硬币个数 。如果没有任何一种硬币组合能组成总金额，返回 -1 。

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)

	for i, _ := range dp {
		dp[i] = math.MaxInt32
	}

	dp[0] = 0
	for i := 1; i < amount; i++ {
		for _, v := range coins {
			if i >= v {
				dp[i] = min(dp[i], dp[i-v]+1)
			}
		}
	}
	if dp[amount] == math.MaxInt32 {
		return -1
	}
	return dp[amount]
}
